waterwheel/wheel_util.py

```diff
@@ -0,0 +1,189 @@
+            matcher.add(key.upper(), None, *phrases)
+            #matcher.add(key.upper(), phrases)
+    except Exception as e:
+        raise RuntimeError(f'Failed to load water bodies from {doc_bin_file_file}.\nError: {str(e)}')
+    
+    return vocab, wikidata, stop_words, matcher
+
+def write_files(vocab, wikidata, phrase_map):
+    """Writes necessary data to resource files.
+    
+    Parameters
+    ----------
+    vocab: Dict
+        A dictionary containing the different water body types and their hash value.
+    wikidata: Dict
+        A dictionary with each wikilink data for every water_body
+    phrase_map: Dict
+        A map of different water body types and DocBin of phrases.
+    """
+    
+    try:
+        with open(vocab_file, 'w') as file:
+            json.dump(vocab, file)
+    except Exception as e:
+        raise RuntimeError(f'Failed to write vocab to {vocab_file}.\nError: {str(e)}')
+    
+    try:
+        with open(wikidata_file, 'w') as file:
+            json.dump(wikidata, file)
+    except Exception as e:
+        raise RuntimeError(f'Failed to write wikidata to {wikidata_file}.\nError: {str(e)}')
+    
+    try:
+        doc_bins = {}
+        for key in phrase_map:
+            doc_bin = DocBin()
+            for doc in phrase_map[key]:
+                doc_bin.add(doc)
+            doc_bins[key] = doc_bin.to_bytes()
+        with open(doc_bin_file, 'wb') as file:
+            pickle.dump(doc_bins, file)
+    except Exception as e:
+        raise RuntimeError(f'Failed to write water_bodies to {doc_bin_file}.\nError: {str(e)}')       
+
+def add_vocab(water_bodies: Dict, nlp, append: bool = True):
+    """Load new vocab and wikidata.
+    
+    Parameters
+    ----------
+    water_bodies : Dict
+        Dictionary containing the list of new water bodies to be loaded.
+        Format:
+        {
+            "LAKE": [(Name, Wiki_Id), ...],
+            "RIVER": [(Name, Wiki_Id), ...],
+            ...
+        }
+    nlp:
+        spacy nlp object
+    append : bool, optional
+        If True then the new data is appended on top of the previous data.
+        Otherwise old data is done away with.
+    """
+    
+    if append:
+        vocab, wikidata, stop_words, matcher = read_files(nlp)
+    else:
+        matcher = PhraseMatcher(nlp.vocab, attr='LOWER')
+        vocab = {}
+        wikidata = {}
+    
+    phrase_map = {}
+    for key in water_bodies:
+        phrases = [nlp(wb) for wb, _ in tqdm(water_bodies[key], desc=f'Loading {key}(s)')]
+        phrase_map[key] = phrases
+        vocab[key] = nlp.vocab.strings[key]
+        matcher.add(key.upper(), None, *phrases)
+        #matcher.add(key.upper(), phrases)
+        if key not in wikidata:
+            wikidata[key] = {}
+        for name, id in water_bodies[key]:
+            wikidata[key][name.lower()] = id
+    
+    write_files(vocab, wikidata, phrase_map)
+
+def add_vocab_csvs(data_dir: Path, nlp, append: bool = True):
```

Comment:
could we also add the type on `nlp` since it's a spacy nlp object? should be possible right?

Reply:
Done.